password_security/models/res_users.py

```diff
@@ -76,19 +52,41 @@ def _check_password_rules(self, password):
         if not password:
             return True
         company_id = self.company_id
-        password_regex = ['^']
-        if company_id.password_lower:
-            password_regex.append('(?=.*?[a-z])')
-        if company_id.password_upper:
-            password_regex.append('(?=.*?[A-Z])')
-        if company_id.password_numeric:
-            password_regex.append(r'(?=.*?\d)')
-        if company_id.password_special:
-            password_regex.append(r'(?=.*?\W)')
-        password_regex.append('.{%d,}$' % company_id.password_length)
-        if not re.search(''.join(password_regex), password):
-            raise PassError(self.password_match_message())
-        return True
+        message = []
+        if company_id.password_lower and sum(map(str.islower, password)) < \
+                company_id.password_lower:
+            message.append('\n ' + _('Lowercase letter (At least ' +
+                                     str(company_id.password_lower) +
+                                     ' character)')
+                           )
+        if company_id.password_upper and sum(map(str.isupper, password)) < \
+                company_id.password_upper:
+            message.append('\n ' + _('Uppercase letter (At least ' +
+                                     str(company_id.password_upper) +
+                                     ' character)')
+                           )
+        if company_id.password_numeric and sum(map(str.isdigit, password)) < \
+                company_id.password_numeric:
+            message.append('\n ' + _('Numeric digit (At least ' +
+                                     str(company_id.password_numeric) +
+                                     ' numeric)')
+                           )
+        if company_id.password_special and len(set('[~!@#$%^&*()_+{}":;\']+$'
```

Comment:
This does not include all special characters. For instance these are missing: `^\/|·,.-`

Simplify and improve with: `if len(re.findall(r"\W", password)) < company_id.password_special`